include/ensmallen_bits/moead/moead_impl.hpp

```diff
@@ -0,0 +1,423 @@
+                    const double lowerBound,
+                    const double upperBound) :
+    populationSize(populationSize),
+    numGeneration(numGeneration),
+    crossoverProb(crossoverProb),
+    mutationProb(mutationProb),
+    mutationStrength(mutationStrength),
+    neighbourhoodSize(neighbourhoodSize),
+    distributionIndex(distributionIndex),
+    lowerBound(lowerBound * arma::ones(1, 1)),
+    upperBound(upperBound * arma::ones(1, 1)),
+    numObjectives(0)
+  { /* Nothing to do here. */ }
+
+//! Optimize the function.
+template<typename MatType,
+         typename... ArbitraryFunctionType,
+         typename... CallbackTypes>
+typename MatType::elem_type MOEAD::Optimize(std::tuple<ArbitraryFunctionType...>& objectives,
+                                            MatType& iterate,
+                                            CallbackTypes&&... callbacks)
+{
+  // Check if lower bound is a vector of a single dimension.
+  if (lowerBound.n_rows == 1)
+    lowerBound = lowerBound(0, 0) * arma::ones(iterate.n_rows, iterate.n_cols);
+
+  // Check if lower bound is a vector of a single dimension.
+  if (upperBound.n_rows == 1)
+    upperBound = upperBound(0, 0) * arma::ones(iterate.n_rows, iterate.n_cols);
+  // Convenience typedefs.
+  typedef typename MatType::elem_type ElemType;
+
+  if(lowerBound.size() != upperBound.size())
+  {
+    throw std::logic_error("MOEAD::Optimize(): size of lowerBound and upperBound "
+        "must be equal.");
+  }
+  if(lowerBound.size() != iterate.n_elem)
+  {
+    throw std::logic_error("MOEAD::Optimize(): there should be a lower bound and "
+        "an upper bound for each variable in the initial point.");
+  }
+  // Number of objective functions. Represented by m in the paper.
+  numObjectives = sizeof...(ArbitraryFunctionType);
+
+  // Controls early termination of the optimization process.
+  bool terminate = false;
+
+  // 1.1 The external population, non-dominated solutions.
+  std::vector<MatType> externalPopulation;
+  std::vector<arma::vec> externalPopulationFValue;
+  std::vector<size_t> shuffle(populationSize);
+
+  // Weight vectors, where each one of them represents a decomposition.
+  std::vector<arma::vec> weights(populationSize);
+  for (size_t i = 0; i < populationSize; i++)
+  {
+    weights[i] = arma::vec(numObjectives, arma::fill::randu);
+    shuffle[i] = i;
+  }
+
+  // 1.2 Storing the indices of nearest neighbours of each weight vector.
+  arma::Mat<size_t> weightNeighbourIndices(populationSize, neighbourhoodSize);
+  for (size_t i = 0; i < populationSize; i++)
+  {
+    // To temporarily store the distance between weights(i) and each other weights.
+    arma::vec distances(populationSize);
+    for (size_t j = 0; j < populationSize; j++)
+    {
+      distances(j) = 0;
+      for (size_t w = 0; w < numObjectives ; w++)
+      {
+        distances(j) += std::pow(weights[i](w) - weights[j](w), 2);
+      }
+      distances(j) = std::sqrt(distances(j));
```

Comment:
Once `weights` is of type `arma::Mat` we can simplify this using `arma::pow, arma::accu` and `.each_col()`, I can elaborate on the strategy here, if needed.